llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -58919,6 +58919,12 @@ bool X86TargetLowering::IsDesirableToPromoteOp(SDValue Op, EVT &PVT) const {
     if (IsFoldableAtomicRMW(N0, Op) ||
         (Commute && IsFoldableAtomicRMW(N1, Op)))
       return false;
+    // When ZU is enabled, we prefer to not promote for MUL by a constant,
```

Comment:
Why constant matters here?

Reply:
IMULZU only supports immediate operands. Other IMUL variants use NDD encodings and aren't covered by this change.

Reply:
I see. So we don't need to consider NDD to non-NDD folding problem here.

Is this already covered by https://github.com/llvm/llvm-project/commit/20683de70e43fa73536ac1e8ce4082604048d040, or you will implement it as a follow up.

Reply:
That commit doesn't touch MUL handling, and I wasn't preparing a follow-up. We should circle back on how to deal with NDD for MUL first.

Reply:
I'm ok with it.